### DISCUSSION TITLE
WSL2 port forwarding is not working if accessed from java

### DISCUSSION THREAD

**Discussion by vigneshr6** [microsoft/WSL#6253] [General]
# Environment
Windows 10 Home
```none
Windows build number: [Version 10.0.19042.630]
Your Distribution version: [Ubuntu 18.04]
Whether the issue is on WSL 2 and/or WSL 1: [WSL 2]
```

# Steps to reproduce
1) Start a webserver in WSL2 ubuntu 18.04
2) Access the webserver from a java HTTP client application from windows.

![image](https://user-images.githubusercontent.com/20204734/99671083-8b464a80-2a97-11eb-8e2f-20eaf63a70a9.png)

**WSL logs**: 

#  Expected behavior
WSL webserver should be accessible with localhost from java application.

# Actual behavior
could not connect the webserver with localhost. But I could connect with private IP from the same java application. Powershell `Invoke-WebRequest` command can also work with localhost.

**Comment by hb020** [comment]
See #4150
This has been discussed often, and MS does not seem to care about it. This behaviour is as designed.
My suggestion: go back to WSL or set up a real VM.

**Comment by therealkenc** [comment]
> 10.0.19042.630

Likely #4851 with a side of #5298, like #5706 ([message](https://github.com/microsoft/WSL/issues/5706#issuecomment-668170117)), others.

Moving over to discussion (similar #6240) for work-arounds to the specific use-case since there isn't a novel actionable.

**Comment by wendersonsouza** [comment]
I got the same issue trying to access a jasperreport server hosted by wsl .
got connection refused attempting  to connect the server from a java application using wsl2.
lead by this tutorial  I set up to wsl1 and then worked well so far.
https://www.tenforums.com/tutorials/164318-how-set-linux-distribution-version-wsl-1-wsl-2-windows-10-a.html

**Comment by albertopiva-tv** [comment]
This issue occurred to me today: a spring boot app launched from WSL2 was not reachable by another spring boot app launched from outside WSL when using PC hostname or localhost. I had to use IP address, but it's not really convenient, it could change.
The issue happens the other way around also, a java app launched from windows is not reachable from WSL using hostname or localhost

**Comment by michaldo** [comment]
In my case connection from Win to WSL2 by Java process depends on property `java.net.preferIPv6Addresses`

Connection works for values **system** and **true**  but does not work for **false** and when missing, because false is default

See https://stackoverflow.com/a/71967275 for details

**Reply by marcoblos** [reply]
After hours trying to figure out possible problems between WSL2 + Windows + Firewall, etc, @michaldo answer was the only one that worked here... Instead of running my application like: `mvn quarkus:dev `, I ran it like this: `mvn quarkus:dev -Djava.net.preferIPv6Addresses=true`.

**Comment by namedgraph** [comment]
This is what helped me to connect VisualVM on Win10 over JMX to a Docker container running on WSL2: https://stackoverflow.com/a/71881475/1003113
